glean-core/src/metrics/timing_distribution.rs

```diff
@@ -0,0 +1,273 @@
+    ///   same timespan metric.
+    /// * `stop_time` - Timestamp in nanoseconds.
+    pub fn set_stop_and_accumulate(&mut self, glean: &Glean, id: TimerId, stop_time: u64) {
+        let duration = match self.timings.set_stop(id, stop_time) {
+            Err(error) => {
+                record_error(glean, &self.meta, ErrorType::InvalidValue, error);
+                return;
+            }
+            Ok(duration) => duration,
+        };
+        let duration = Duration::from_nanos(duration);
+        let duration = self.time_unit.duration_convert(duration);
+
+        if !self.should_record(glean) {
+            return;
+        }
+
+        glean
+            .storage()
+            .record_with(&self.meta, |old_value| match old_value {
+                Some(Metric::TimingDistribution(mut hist, time_unit)) => {
+                    hist.accumulate(duration as u32);
+                    Metric::TimingDistribution(hist, time_unit)
+                }
+                _ => {
+                    let mut hist = Histogram::default();
+                    hist.accumulate(duration as u32);
+                    Metric::TimingDistribution(hist, self.time_unit)
+                }
+            });
+    }
+
+    /// Abort a previous [start] call. No error is recorded if no [start] was called.
+    ///
+    /// @param timerId The [GleanTimerId] to associate with this timing. This allows
+    /// for concurrent timing of events associated with different ids to the
+    /// same timing distribution metric.
+    ///
+    pub fn cancel(&mut self, id: TimerId) {
+        self.timings.cancel(id);
+    }
+
+    /// **Test-only API (exported for FFI purposes).**
+    ///
+    /// Get the currently stored value as an integer.
+    ///
+    /// This doesn't clear the stored value.
+    pub fn test_get_value(&self, glean: &Glean, storage_name: &str) -> Option<Histogram> {
+        match StorageManager.snapshot_metric(glean.storage(), storage_name, &self.meta.identifier())
+        {
+            Some(Metric::TimingDistribution(hist, _)) => Some(hist),
+            _ => None,
+        }
+    }
+
+    /// **Test-only API (exported for FFI purposes).**
+    ///
+    /// Get the currently-stored histogram as a JSON String of the serialized value.
+    ///
+    /// This doesn't clear the stored value.
+    pub fn test_get_value_as_json_string(
+        &self,
+        glean: &Glean,
+        storage_name: &str,
+    ) -> Option<String> {
+        self.test_get_value(glean, storage_name)
+            .map(|hist| serde_json::to_string(&snapshot(&hist, self.time_unit)).unwrap())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn can_snapshot() {
+        use serde_json::json;
+
+        let mut hist = Histogram::exponential(1, 500, 10);
+
+        for i in 1..=10 {
+            hist.accumulate(i);
+        }
+
+        let snap = snapshot(&hist, TimeUnit::Millisecond);
+
+        // TODO: Do we need to trim empty buckets?
```

Comment:
Do we have a bug filed for this? If not, can you file it? If yes, can you add a reference to the bug number here?

Reply:
We trim. Removing todo.